docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,115 @@
+# Introduction
+There is [external health monitoring](https://github.com/kubernetes-csi/external-health-monitor/) which does mostly similar stuff, but actions are not in the scope of it, all it can do is log the health condition of the mounts.
+
+# Volume Healer Controller
+
+**Now the obvious question we expect is "what do we achieve with the new controller?"**
+* Achieve a way to monitor the health of volumes.
+* Ceph-CSI has healer component which take cares of restarting the userspace processes when the node plugin is restarted, but in many places it uses Kubenetes specific code, with this controller model we are moving out the Kubernetes specific code out, and help make Ceph-CSI orchestrator independent.
+* We currently do not have a way to act on failed volumes. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it. For example userspace mounters such as NBD has a process running per volume, this controller should ideally help restart such processes in case if they are not-started or crashed.
+* This will be a generic controller design for any mounters added in the future, for userspace mounters it will offer healing in addition, like restarting the respective processes if the volume goes abnormal. Some example userspace mounters would be rbd-nbd, cephfs-fuse, etc.
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller. It will trigger controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Fetch all the VolumeAttachment` list
+    * Filter the `VolumeAttachment` list through matching driver name, node name and status attached
+    * For each `VolumeAttachment` get the respective PersistentVolume information and check the criteria of PersistentVolume Bound (also we can have configuration options like mounter type)
```

Comment:
Send request irrespective of any particular PV config option ?
Or
let the driver include more information in its capabilities response, so this decision can be made with the help of the extra info provided by the driver itself ?

Reply:
My idea is similar to the latter one, somehow get the choice like, what kind of mounter types need what services.
Example:
```
{
  rbd-nbd: healing
  krbd: monitoring
  xyz: disable
}
```

Reply:
I would like to leave this more for an implementation-specific detail.

Reply:
CSI-Addons capabilities from a CSI-driver should include if it supports healing a volume. If the CSI-driver does not need healing, it can omit the capability.
In case the CSI-driver needs to perform heal checks on a volume (depending on the volume options?), it can either return things like `HEALTH_OK` (to recheck later), `HEALTH_UNSUPPORTED` (no need to check, this volume does not support healing), `HEALTH_BAD` (needs healing) and so on. But this is a discussion for csi-addons/spec.

Reply:
CSI-Addons should not need to know about different _mounter types_, that would require extending the list of types and the implementation for each CSI-driver. Better have a generic approach that works for all CSI-drivers, so that any CSI-driver can opt-in without changing CSI-Addons itself.

Reply:
This sounds reasonable.

`HEALTH_OK`, `HEALTH_UNSUPPORTED`, `HEALTH_BAD` are some kind of volume attributes that CSI driver sets on the persistent volume?

In my previous comment, I mean something like:
* capability response returns some string that holds
```
{
  Mounter1: NEED_HEALING
  Mounter2: NEED_MONITORING
  Mounter3: UNKNOWN
}
```
* If the PV attribute mounter matches with `Mounter1`, attempt to heal it.
* If it is of type `Mounter2` only check the health,
* If it is of type `Mounter3` do not attempt anything from the healer controller.

Reply:
Interactions between CSI-Addons and the CSI-driver need to be documented in a [csi-addons/spec](https://github.com/csi-addons/spec). The protocol then describes what a CSI-driver needs to implement for supporting the healer operations. This will probably include a description for capabilities of the CSI-driver. Certain operations on a volume can return attributes that healing is not supported for the volume, but other volumes from the same driver might support more features.

I do not think CSI-Addons should know about internals of the CSI-driver, so _mounters_ would not be needed,